File: problem68/Problem68.java
import java.util.*;
/**
 * Problem 68
 */

public class Problem68 {
	public static String max = "0";
	
	public static void main(String[] arg) {
		int n = 5; // n-gon ring
		// Get couples
		ArrayList<ArrayList<Integer>> couples = new ArrayList<ArrayList<Integer>>();
		for (int i = 1; i < 2 * n + 1; i++) {
			for (int j = i + 1; j < 2 * n + 1; j++) {
				ArrayList<Integer> coupleOne = new ArrayList<Integer>();
				ArrayList<Integer> coupleTwo = new ArrayList<Integer>();
				coupleOne.add(i); coupleOne.add(j);
				coupleTwo.add(j); coupleTwo.add(i);
				couples.add(coupleOne); couples.add(coupleTwo);
			}
		}
		boolean[] legal = new boolean[couples.size()];
		for (int i = 0; i < legal.length; i++) {
			legal[i] = true;
		}
		ArrayList<Integer> solution = new ArrayList<Integer>();
		solve(solution, couples, legal, n);
		System.out.println("The solution is " + max);
	}
	
	/**
	* Returns a solution set given a solution expressed in terms of indices of elements in an array of couples
	*/
	public static int[][] solutionSet(ArrayList<Integer> solution, ArrayList<ArrayList<Integer>> couples) {
		int[][] result = new int[solution.size()][3];
		for (int i = 0; i < solution.size(); i++) {
			for (int j = 0; j < couples.get(solution.get(i)).size(); j++) {
				result[i][j] = couples.get(solution.get(i)).get(j);
			}
			result[i][couples.get(i).size()] = couples.get(solution.get((i + 1) % solution.size())).get(1);
		}
		return result;
	}
	
	/**
	* Returns if a > b where a and b are integers represented as strings
	*/
	public static boolean greaterThan(String a, String b) {
		if (a.length() != b.length()) return a.length() > b.length();
		for (int i = 0; i < a.length(); i++) {
			int digitA = Integer.parseInt(a.substring(i, i + 1));
			int digitB = Integer.parseInt(b.substring(i, i + 1));
			if (digitA != digitB) return digitA > digitB;
		}
		return false;
	}
	
	/**
	* Converts a solution set into a string
	*/
	public static String solutionSetAsStr(int[][] solutionSet) {
		String result = "";
		for (int i = 0; i < solutionSet.length; i++) {
			for (int j = 0; j < solutionSet[i].length; j++) {
				result += Integer.toString(solutionSet[i][j]);
			}
		}
		return result;
	}
	
	/**
	* Recursive backtracking solver
	*/
	public static void solve(ArrayList<Integer> solution, ArrayList<ArrayList<Integer>> couples, boolean[] legal, int n) {
		if (solution.size() == n) {
			int total = couples.get(solution.get(0)).get(0) + couples.get(solution.get(0)).get(1) + couples.get(solution.get(1)).get(1);
			if (couples.get(solution.get(solution.size() - 1)).get(0) + couples.get(solution.get(solution.size() - 1)).get(1) + couples.get(solution.get(0)).get(1) == total) {
				int[][] solSet = solutionSet(solution, couples);
				String solSetAsStr = solutionSetAsStr(solSet);
				if (greaterThan(solSetAsStr, max) && solSetAsStr.length() == 16) max = solSetAsStr;
				return;
			}
		}
		else {
			if (solution.size() == 0) {
				for (int i = 0; i < couples.size(); i++) {
					solution.add(i);
					boolean[] sievedLegal = sieveCouples2(couples.get(i).get(0), couples, sieveCouples1(couples.get(i).get(0), couples.get(i).get(1), couples, legal));
					solve(copy(solution), couples, sievedLegal, n);
					solution.remove(solution.size() - 1);
				}
			}
			if (solution.size() == 1) {
				for (int i = 0; i < couples.size(); i++) {
					if (legal[i]) {
						if (couples.get(i).get(0) < couples.get(solution.get(0)).get(0) + couples.get(solution.get(0)).get(1)) {
							solution.add(i);
							boolean[] sievedLegal = sieveCouples1(couples.get(i).get(0), couples.get(i).get(1), couples, legal);
							solve(copy(solution), couples, sievedLegal, n);
							solution.remove(solution.size() - 1);
						}
					}
				}
			}
			if (solution.size() > 1) {
				for (int i = 0; i < couples.size(); i++) {
					if (legal[i]) {
						int total = couples.get(solution.get(0)).get(0) + couples.get(solution.get(0)).get(1) + couples.get(solution.get(1)).get(1);
						if (couples.get(i).get(1) + couples.get(solution.get(solution.size() - 1)).get(0) + couples.get(solution.get(solution.size() - 1)).get(1) == total) {
							solution.add(i);
							boolean[] sievedLegal = sieveCouples1(couples.get(i).get(0), couples.get(i).get(1), couples, legal);
							solve(copy(solution), couples, sievedLegal, n);
							solution.remove(solution.size() - 1);
						}
					}
				}
			}
		}
	}

	/**
	* Returns an updated array of booleans
	*/
	public static boolean[] sieveCouples1(int a, int b, ArrayList<ArrayList<Integer>> couples, boolean[] T) {
		boolean[] result = new boolean[T.length];
		for (int i = 0; i < result.length; i++) {
			if (couples.get(i).indexOf(a) > -1 || couples.get(i).indexOf(b) > -1) result[i] = false;
			else result[i] = T[i];
		}
		return result;
	}
	
	/**
	* Returns an updated array of booleans
	*/
	public static boolean[] sieveCouples2(int a, ArrayList<ArrayList<Integer>> couples, boolean[] T) {
		boolean[] result = new boolean[T.length];
		for (int i = 0; i < couples.size(); i++) {
			if (couples.get(i).get(0) <= a) result[i] = false;
			else result[i] = T[i];
		}
		return result;
	}
	
	/**
	* Deep copies a dynamic array 
	*/
	public static ArrayList<Integer> copy(ArrayList<Integer> T) {
		ArrayList<Integer> result = new ArrayList<Integer>();
		for (int i = 0; i < T.size(); i++) {
			result.add(T.get(i));
		}
		return result;
	}
}

File: problem68/problem68.js
/**
 * Problem 68
 */

var max = '0';
var n = 5; // n-gon ring
// Get couples
couples = [];
for (var i = 1; i < 2 * n + 1; i++) {
	for (var j = i + 1; j < 2 * n + 1; j++) {
		var coupleOne = [];
		var coupleTwo = [];
		coupleOne.push(i); coupleOne.push(j);
		coupleTwo.push(j); coupleTwo.push(i);
		couples.push(coupleOne); couples.push(coupleTwo);
	}
}
legal = [];
for (var i = 0; i < couples.length; i++) {
	legal.push(true);
}
var solution = [];
solve(solution, couples, legal, n);
console.log('The solution is ' + max);

/**
* Returns a solution set given a solution expressed in terms of indices of elements in an array of couples
* @param array of integers solution, two dimensional array of integers couples
* @return two dimensional array of integers
*/

function solutionSet(solution, couples) {
	var result = []; 
	for (var i = 0; i < solution.length; i++) {
		var row = [];
		for (var j = 0; j < 3; j++) {
			row.push(0);
		}
		result.push(row);
	}
	for (var i = 0; i < solution.length; i++) {
		for (var j = 0; j < couples[solution[i]].length; j++) {
			result[i][j] = couples[solution[i]][j];
		}
		result[i][couples[i].length] = couples[solution[(i + 1) % solution.length]][1];
	}
	return result;
}

/**
* Returns if a > b where a and b are integers represented as strings
* @param string a, string b
* @return boolean
*/
function greaterThan(a, b) {
	if (a.length != b.length) return a.length > b.length;
	for (var i = 0; i < a.length; i++) {
		var digitA = parseInt(a.substring(i, i + 1));
		var digitB = parseInt(b.substring(i, i + 1));
		if (digitA != digitB) return digitA > digitB;
	}
	return false;
}

/**
* Converts a solution set into a string
* @param two dimensional array of integers solutionSet
* @return string
*/
function solutionSetAsStr(solutionSet) {
	var result = '';
	for (var i = 0; i < solutionSet.length; i++) {
		for (var j = 0; j < solutionSet[i].length; j++) {
			result += solutionSet[i][j].toString();
		}
	}
	return result;
}

/**
* Recursive backtracking solver
* @param array of integers solution, two dimensional array of integers couples, array of booleans legal, integer n
* @return void
*/
function solve(solution, couples, legal, n) {
	if (solution.length == n) {
		var total = couples[solution[0]][0] + couples[solution[0]][1] + couples[solution[1]][1];
		if (couples[solution[solution.length - 1]][0] + couples[solution[solution.length - 1]][1] + couples[solution[0]][1] == total) {
			var solSet = solutionSet(solution, couples);
			var solSetAsStr = solutionSetAsStr(solSet);
			if (greaterThan(solSetAsStr, max) && solSetAsStr.length == 16) max = solSetAsStr;
			return;
		}
	}
	else {
		if (solution.length == 0) {
			for (var i = 0; i < couples.length; i++) {
				solution.push(i);
				var sievedLegal = sieveCouples2(couples[i][0], couples, sieveCouples1(couples[i][0], couples[i][1], couples, legal));
				solve(copy(solution), couples, sievedLegal, n);
				solution.splice(solution.length - 1, 1);
			}
		}
		if (solution.length == 1) {
			for (var i = 0; i < couples.length; i++) {
				if (legal[i]) {
					if (couples[i][0] < couples[solution[0]][0] + couples[solution[0]][1]) {
						solution.push(i);
						var sievedLegal = sieveCouples1(couples[i][0], couples[i][1], couples, legal);
						solve(copy(solution), couples, sievedLegal, n);
						solution.splice(solution.length - 1, 1);
					}
				}
			}
		}
		if (solution.length > 1) {
			for (var i = 0; i < couples.length; i++) {
				if (legal[i]) {
					var total = couples[solution[0]][0] + couples[solution[0]][1] + couples[solution[1]][1];
					if (couples[i][1] + couples[solution[solution.length - 1]][0] + couples[solution[solution.length - 1]][1] == total) {
						solution.push(i);
						var sievedLegal = sieveCouples1(couples[i][0], couples[i][1], couples, legal);
						solve(copy(solution), couples, sievedLegal, n);
						solution.splice(solution.length - 1, 1);
					}
				}
			}
		}
	}
}

/**
* Returns an updated array of booleans
* @param integer a, integer b, two dimensional array of integers couples, array of booleans T
* @return array of booleans
*/
function sieveCouples1(a, b, couples, T) {
	var result = [];
	for (var i = 0; i < T.length; i++) {
		if (couples[i].indexOf(a) > -1 || couples[i].indexOf(b) > -1) result.push(false);
		else result.push(T[i]);
	}
	return result;
}

/**
* Returns an updated array of booleans
* @param integer a, two dimensional array of integers couples, array of booleans T
* @return array of booleans
*/
function sieveCouples2(a, couples, T) {
	var result = [];
	for (var i = 0; i < couples.length; i++) {
		if (couples[i][0] <= a) result.push(false);
		else result.push(T[i]);
	}
	return result;
}

/**
* Deep copies of a dynamic array
* @param array of integers T
* @return array of integers 
*/
function copy(T) {
	var result = [];
	for (var i = 0; i < T.length; i++) {
		result.push(T[i]);
	}
	return result;
}